app/src/main/java/pm/gnosis/heimdall/views/StepIndicator.kt

```diff
@@ -0,0 +1,331 @@
+
+    private val backgroundPaint: Paint
+
+    init {
+        orientation = HORIZONTAL
+        gravity = Gravity.CENTER
+        backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+        backgroundPaint.style = Paint.Style.FILL_AND_STROKE
+        backgroundPaint.strokeWidth = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2f, resources.displayMetrics)
+    }
+
+    fun updateStep(index: Int, stepState: StepState) {
+        val step = getChildAt(index) as Step
+        step.state = stepState
+        invalidate()
+    }
+
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        if (changed) {
+            var childsMargin = if (childCount == 0) 0 else (width - paddingLeft * 2 - getChildAt(0).measuredWidth * childCount) / childCount
+            var left = l + paddingLeft + childsMargin / 2
+            for (i in 0 until childCount) {
+                val child = getChildAt(i)
+
+                val childWidth = child.measuredWidth
+                val childHeight = child.measuredHeight
+
+                child.layout(left + childsMargin / 2, paddingTop, left + childWidth + childsMargin / 2, childHeight + paddingTop)
+                left += childWidth + childsMargin / 2
+            }
+        }
+    }
+
+    override fun dispatchDraw(canvas: Canvas) {
+        super.dispatchDraw(canvas)
+
+        for (i in 0 until childCount - 1) {
+
+            val child1 = getChildAt(i) as Step
+            val child2 = getChildAt(i + 1) as Step
+
+            val colorPair =
+                if (child1.state == StepState.UNCOMPLETED_INACTIVE || child1.state == StepState.SKIPPED) {
+
+                    if (child2.state == StepState.UNCOMPLETED_INACTIVE || child2.state == StepState.SKIPPED)
+                        Pair(
+                            ContextCompat.getColor(context, R.color.disabled_button),
+                            ContextCompat.getColor(context, R.color.disabled_button)
+                        )
+                    else
+                        Pair(
+                            ContextCompat.getColor(context, R.color.disabled_button),
+                            ContextCompat.getColor(context, R.color.safe_green)
+                        )
+
+                } else {
+
+                    if (child2.state == StepState.UNCOMPLETED_INACTIVE || child2.state == StepState.SKIPPED)
+                        Pair(
+                            ContextCompat.getColor(context, R.color.safe_green),
+                            ContextCompat.getColor(context, R.color.disabled_button)
+                        )
+                    else
+                        Pair(
+                            ContextCompat.getColor(context, R.color.safe_green),
+                            ContextCompat.getColor(context, R.color.safe_green)
+                        )
+                }
+
+
+            val shader = LinearGradient(
+                child1.getCircleCenter().x + child1.radius(),
```

Comment:
Would probably be nice to somehow cache this, since you are creating an object each time you call this, which is a bad practice in `onDraw`

Reply:
https://developer.android.com/training/custom-views/custom-drawing#createobject

Reply:
Same actually for the color pair